### PR TITLE
feat!: allow disabling notifications

### LINKED ISSUE #15513: Support disabling notifications
### Problem Description

Not all coder deployments will want notifications enabled, so it would be nice to be able to disable the feature. Currently, if you leave the notification settings as default, there are error logs that appear every time the system attempts to dispatch a notification.

### Desired Solution

Add a flag `CODER_NOTIFICATIONS_ENABLED` that defaults to `true` but can be set to `false` to disable notifications.

### PR DESCRIPTION
**Pull request by DanielleMaywood**
Resolves https://github.com/coder/coder/issues/15513

~~Adds a new option (`$CODER_NOTIFICATIONS_ENABLED`) that allows disabling notifications. This defaults to `true` to avoid disabling notifications for existing installations of Coder.~~

Disables notifications when both `$CODER_NOTIFICATIONS_WEBHOOK_ENDPOINT` and `$CODER_EMAIL_SMARTHOST` are unset.

Drops the default value for `$CODER_EMAIL_SMARTHOST`.

### CODE REVIEW

File: codersdk/deployment_test.go [hunk 1]
```diff
@@ -568,3 +568,115 @@ func TestPremiumSuperSet(t *testing.T) {
 	require.NotContains(t, enterprise.Features(), "", "enterprise should not contain empty string")
 	require.NotContains(t, premium.Features(), "", "premium should not contain empty string")
 }
+
+func TestEmailValuesTakeCorrectPrecedent(t *testing.T) {
```
Comment (dannykopping) [inline codersdk/deployment_test.go:572]: Non-blocking nit: this feels out of place in this PR; perhaps we should land your previous PR (https://github.com/coder/coder/pull/15482) with just the tests rather than muddying this PR with unrelated changes.

File: codersdk/deployment.go [hunk 2]
```diff
@@ -2595,22 +2597,17 @@ Write out the current server config as YAML to stdout.`,
...
-		emailTLSSkipCertVerify,
-		emailTLSCertAuthorityFile,
-		emailTLSCertFile,
-		emailTLSCertKeyFile,
 		// Notifications Options
+		{
+			Name:        "Notifications: Enabled",
+			Description: "Controls if notifications are enabled.",
+			Flag:        "notifications-enabled",
+			Env:         "CODER_NOTIFICATIONS_ENABLED",
```
Comment (dannykopping) [inline codersdk/deployment.go:2605]: Can you please add some tests validating this functionality?
For example, we need to ensure enqueuing notifications doesn't cause issues when notifications are disabled, etc (although I don't see anything in this PR relating to that - which is a smell).

File: codersdk/deployment_test.go (hunk 1 above)
Comment (DanielleMaywood) [reply codersdk/deployment_test.go:572]: Sure thing, I'll back these changes out from this PR.

File: codersdk/deployment.go (hunk 2 above)
Comment (DanielleMaywood) [reply codersdk/deployment.go:2605]: Sure thing!

This PR does make a small change to the logic of setting the enqueuer

The enqueuer is by default set as the `notifications.NoopEnqueuer`.

https://github.com/coder/coder/blob/c51311310575c9be8894367b4bc0b8e91e764034/cli/server.go#L624

We then re-assign the enqueuer only if notifications are enabled.

https://github.com/coder/coder/blob/c51311310575c9be8894367b4bc0b8e91e764034/cli/server.go#L905-L914

Comment (dannykopping) [reply codersdk/deployment.go:2605]: Ah, you're right! Great 👍

File: codersdk/deployment.go [hunk 3]
```diff
@@ -2871,6 +2868,21 @@ Write out the current server config as YAML to stdout.`,
 			Annotations: serpent.Annotations{}.Mark(annotationFormatDuration, "true"),
 			Hidden:      true, // Hidden because most operators should not need to modify this.
 		},
+		// Email options
+		emailFrom,
```
Comment (dannykopping) [inline codersdk/deployment.go:2872]: You can probably revert these changes, too.

File: cli/server.go [hunk 4]
```diff
@@ -897,31 +897,37 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
 			}
 
 			// Manage notifications.
-			cfg := options.DeploymentValues.Notifications
-			metrics := notifications.NewMetrics(options.PrometheusRegistry)
-			helpers := templateHelpers(options)
+			var (
+				cfg                  = options.DeploymentValues.Notifications
```
Comment (mafredri) [inline cli/server.go:901]: Nit: This variable name is too generic without being limited by function scope.

File: codersdk/deployment.go [hunk 5]
```diff
@@ -2611,6 +2613,16 @@ Write out the current server config as YAML to stdout.`,
 		emailTLSCertFile,
 		emailTLSCertKeyFile,
 		// Notifications Options
+		{
+			Name:        "Notifications: Enabled",
+			Description: "Controls if notifications are enabled.",
```
Comment (mafredri) [inline codersdk/deployment.go:2618]:
```suggestion
			Description: "Enable or disable notifications.",
```
Suggestion: Simplify

**Review by mafredri — APPROVED** [review]
Minor nits but otherwise LGTM 👍🏻

File: cli/server.go (hunk 4 above)
Comment (DanielleMaywood) [reply cli/server.go:901]: This was the existing name, but I'm happy to change it as I agree with you.

**Review by johnstcn — APPROVED** [review]
Nothing blocking from my side!

**Comment by DanielleMaywood** [comment]
I've made some changes to the approach taken here due to feedback in https://github.com/coder/coder/issues/15513. Rather than having a flag that enables/disables notifications, we figure it out based on if certain env variables are set. To do this we are making a breaking change by dropping the default for `CODER_EMAIL_SMARTHOST`, although this shouldn't be a major issue as anyone using SMTP notifications will most likely have changed this setting anyway.

File: cli/server.go [hunk 6]
```diff
@@ -897,31 +897,39 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
...
+				notificationsManager.Run(dbauthz.AsNotifier(ctx))
 
-			// Run report generator to distribute periodic reports.
-			notificationReportGenerator := reports.NewReportGenerator(ctx, logger.Named("notifications.report_generator"), options.Database, options.NotificationsEnqueuer, quartz.NewReal())
-			defer notificationReportGenerator.Close()
+				// Run report generator to distribute periodic reports.
+				notificationReportGenerator := reports.NewReportGenerator(ctx, logger.Named("notifications.report_generator"), options.Database, options.NotificationsEnqueuer, quartz.NewReal())
+				defer notificationReportGenerator.Close()
+			} else {
+				logger.Info(ctx, "notifications disabled as there are no notification methods configured")
```
Comment (dannykopping) [inline cli/server.go:931]: Stylistic nit: errors are lower-cased, log messages initial-cased.

```suggestion
				logger.Info(ctx, "Notifications are currently disabled as there are no notification methods configured.")
```

> as there are no notification methods configured

Putting my tech writer hat on here: we have `CODER_NOTIFICATIONS_METHOD` which has a default value, so this could get confusing. Maybe you can rephrase for clarity.

Also, we should tell the user how to enable notifications in this case. Linking to https://coder.com/docs/admin/monitoring/notifications should suffice.

PS please ensure that docs page is aligned with the changes in this PR.

File: codersdk/deployment.go [hunk 7]
```diff
@@ -686,11 +686,15 @@ type NotificationsConfig struct {
 	Webhook NotificationsWebhookConfig `json:"webhook" typescript:",notnull"`
 }
 
+func (n *NotificationsConfig) Enabled() bool {
+	return n.SMTP.Smarthost != "" || n.Webhook.Endpoint != serpent.URL{}
```
Comment (dannykopping) [inline codersdk/deployment.go:690]: We already have a validation function for smarthost; let's centralise that logic (i.e. check for `ValidationNoSmarthostErr`).

We could do the same for webhook. If both return errors, neither are configured correctly.

**Review by dannykopping — COMMENTED** [review]
Looking good, a couple more things to tweak and we can get this merged today 👍

File: codersdk/deployment.go [hunk 8]
```diff
@@ -686,11 +686,15 @@ type NotificationsConfig struct {
 	Webhook NotificationsWebhookConfig `json:"webhook" typescript:",notnull"`
 }
 
+func (n *NotificationsConfig) Enabled() bool {
```
Comment (dannykopping) [inline codersdk/deployment.go:689]: Can you please add some tests for this specifically?

File: cli/server.go (hunk 6 above)
Comment (dannykopping) [reply cli/server.go:931]: Ah, I thought I read `cliui.Info`. It should probably be that to increase the discoverability of the message.

Comment (DanielleMaywood) [reply cli/server.go:931]: Ah yeah that will probably be better, I will use that instead, thank you.

File: codersdk/deployment_test.go [hunk 9]
```diff
@@ -568,3 +568,69 @@ func TestPremiumSuperSet(t *testing.T) {
...
+					Name:  "CODER_EMAIL_SMARTHOST",
+					Value: "localhost:587",
+				},
+			},
+			expectNotificationsEnabled: true,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
Comment (dannykopping) [inline codersdk/deployment_test.go:622]: BTW this is no longer necessary [as of Go v1.22](https://go.dev/blog/loopvar-preview).

**Review by dannykopping — APPROVED** [review]
LGTM, thanks @DanielleMaywood!

File: codersdk/deployment_test.go (hunk 9 above)
Comment (DanielleMaywood) [reply codersdk/deployment_test.go:622]: Should I add a linter exception for this so it doesn't complain?

Comment (dannykopping) [reply codersdk/deployment_test.go:622]: That'd be wonderful; in a separate PR though - let's land this one

**Comment by johnstcn** [comment]
> I've made some changes to the approach taken here due to feedback in #15513. Rather than having a flag that enables/disables notifications, we figure it out based on if certain env variables are set. To do this we are making a breaking change by dropping the default for `CODER_EMAIL_SMARTHOST`, although this shouldn't be a major issue as anyone using SMTP notifications will most likely have changed this setting anyway.

Gotcha! FYI, if we do a breaking change, we should indicate that in the commit message (e.g. `feat!: ...`

**Review by johnstcn — APPROVED** [review]
